### PR TITLE
Abiquo Compute Driver

### PR DESCRIPTION
**Pull request by devvesa**
Hi!

Finally I finished the Abiquo Compute Driver. Sorry for the delay!!

This is related to jira issue: https://issues.apache.org/jira/browse/LIBCLOUD-250

### CODE REVIEW

File: libcloud/common/abiquo.py [hunk 1]
```diff
@@ -0,0 +1,230 @@
...
+
+    def parse_error(self):
+        """ Parse the error messages
+
+        Response body can easily be handled by this class parent
+        L{XmlResponse}, but there are use cases which Abiquo API
+        does not respond an XML but an HTML. So we need to
+        handle these special cases"""
+        if self.status == httplib.UNAUTHORIZED:
+            raise InvalidCredsError(driver='AbiquoNodeDriver')
```
Comment (Kami) [inline libcloud/common/abiquo.py]: This should be `self.connection.driver` or `self.driver`. It should point to the actual driver instance.

File: libcloud/common/abiquo.py [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+            raise LibcloudError(errors[0].findtext('message'))
+
+    def success(self):
+        """ Determine if the request was successful.
+
+        Any of the 2XX HTTP response codes are accepted as successfull requests
+
+        @rtype:  C{bool}
+        @return: successful request or not"""
+        return (self.status == httplib.OK or self.status == httplib.CREATED or
```
Comment (Kami) [inline libcloud/common/abiquo.py]: Would be a bit shorter and easier to read if you do `return self.status in [httplib.OK, httplib.CREATED, ...]`.

File: libcloud/common/abiquo.py [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        in each request
+
+        @type  headers: C{dict}
+        @param headers: Default input headers
+        @rtype          C{dict}
+        @return:        Default input headers with the 'Authorization'
+                            header"""
+        encoded = base64.b64encode(b('%s:%s' % (self.user_id, self.key)))
+
+        if sys.version_info >= (3, 0):
```
Comment (Kami) [inline libcloud/common/abiquo.py]: I don't think you need this check, you can always just call `.decode('utf-8')` on `encoded` variable.

File: libcloud/common/abiquo.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+    def has_completed(self, response):
+        """ Decide if the asynchronous job has ended
+
+        @type  response: C{xml.etree.ElementTree}
+        @param response: Response object returned by poll request
+        @rtype:          C{bool}
+        @return:         Whether the job has completed"""
+        task = response.object
+        task_state = task.findtext('state')
+        return (task_state == 'FINISHED_SUCCESSFULLY' or
```
Comment (Kami) [inline libcloud/common/abiquo.py]: Same here - `task_state in [..]`

File: libcloud/common/abiquo.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+                task_state == 'FINISHED_UNSUCCESSFULLY' or
+                task_state == 'ABORTED')
+
+
+class ForbiddenError(LibcloudError):
+    """Exception used when credentials are ok but user has not permissions"""
+
+    def __init__(self):
+        message = 'User has not permission to perform this task.'
+        super(LibcloudError, self).__init__(message, 'AbiquoNodeDriver')
```
Comment (Kami) [inline libcloud/common/abiquo.py]: Second argument (driver) needs to point to the actual instance.

File: libcloud/compute/drivers/abiquo.py [hunk 6]
```diff
@@ -0,0 +1,702 @@
...
+        @return:               The newly created node.
+        @rtype:                L{Node}
+        """
+
+        #first, get image location
+        if not 'image' in kwargs:
+            error = "'image' parameter is mandatory"
+            raise LibcloudError(error, self)
+
+        image = kwargs['image']
```
Comment (Kami) [inline libcloud/compute/drivers/abiquo.py]: This function is a bit long and hard to read / follow. Is it possible to split it in multiple smaller functions (e.g. _find_image, _register_group, ...)?

File: libcloud/test/compute/test_abiquo.py [hunk 7]
```diff
@@ -0,0 +1,401 @@
...
+        self.driver = AbiquoNodeDriver('son', 'goku',
+                                       'http://dummy.host.com/api')
+
+    def test_unauthorized_controlled(self):
+        """Test the Unauthorized Exception is Controlled
+
+        Test, through the 'login' method, that a '401 Unauthorized'
+        raises a 'InvalidCredsError' instead of the 'MalformedUrlException'
+        """
+        AbiquoNodeDriver.connectionCls.conn_classes = (AbiquoMockHttp, None)
```
Comment (Kami) [inline libcloud/test/compute/test_abiquo.py]: You don't need to set `conn_classes` again here, because you already set it in `setUp`.

**Comment by Kami** [comment]
@jdevesa Looks better, thanks.

Can you please attach final patch to the JIRA ticket so I can do a final review and merge it?

**Comment by devvesa** [comment]
Done!

**Pull request closed without merging**